chainer/training/extension.py

```diff
@@ -0,0 +1,109 @@
+PRIORITY_WRITER = 300
+PRIORITY_EDITOR = 200
+PRIORITY_READER = 100
+
+
+class Extension(object):
+
+    """Base class of trainer extensions.
+
+    Extension of :class:`Trainer` is a callable object that takes the trainer
+    object as the argument. It also provides some default configurations as its
+    attributes, e.g. the default trigger and the default priority. This class
+    provides a set of typical default values for these attributes.
+
+    There are two ways to define users' own extensions: inheriting this class,
+    or decorating closures by :func:`make_extension`. Decorator can slightly
+    reduce the overhead and is much easier to use, while this class provides
+    more flexibility (for example, it can have methods to configure the
+    behavior).
+
+    Attributes:
+        trigger: Default value of trigger for this extension. It is set to
+            ``(1, 'iteration')`` by default.
+        priority: Default priority of the extension. It is set to
+            ``PRIORITY_READER`` by default.
+        invoke_before_training: Default flag to decide whether this extension
+            should be invoked before the training starts. The default value is
+            ``False``.
+
+    """
+    trigger = 1, 'iteration'
+    priority = PRIORITY_READER
+    invoke_before_training = False
+
+    @property
+    def default_name(self):
+        """Default name of the extension.
+
+        It is the name of the class by default. Implementation can override
+        this property, or provide a class attribute to hide it.
+
+        """
+        return type(self).__name__
+
+    def __call__(self, trainer):
+        """Invokes the extension.
+
+        Implementations should override this operator. This method is called
+        at iterations which the corresponding trigger accepts.
+
+        Args:
+            trainer (Trainer): Trainer object that calls this operator.
+
+        """
+        pass
+
+    def finalize(self):
+        """Finalizes the extension.
+
+        This method is called at the end of the training loop.
+
+        """
+        pass
+
+    def serialize(self, serializer):
+        """Serializes the extension state.
+
+        It is called when a trainer that owns this extension is serialized. It
+        serializes nothing by default.
+
+        """
+        pass
```

Comment:
The base class `Updater` raises NotImplementError in `serialize` method, while `Extension` class does not. Is it intentional?